src/main/java/com/cryptomorin/xseries/particles/ParticleDisplay.java

```diff
@@ -1794,4 +1833,221 @@ public Vector mul(Vector point) {
             return new Vector(vx, vy, vz);
         }
     }
+
+    public interface ParticleData {
+        /**
+         * Returns true if the data type uses the offset fields,
+         * and false if it uses the data field.
+         * Sometimes depends on the particle and version.
+         */
+        public default boolean isOffsetBased(XParticle particle) {
+            return false;
+        }
```

Comment:
I believe you should just return null for `offsetValues()` if it's not supported. This is kinda extra.